src/ray/raylet/node_manager.cc

```diff
@@ -1933,79 +1933,127 @@ void NodeManager::FinishAssignedActorTask(Worker &worker, const Task &task) {
+          }
+        },
+        //failure_callback
+        [this, task, actor_id, actor_handle_id, new_actor_data, resumed_from_checkpoint]
+            (ray::gcs::AsyncGcsClient *client, const TaskID &parent_task_id) mutable {
+          // The parent task was not in the GCS task table. It must therefore be in the
+          // lineage cache.
+          RAY_CHECK(lineage_cache_.ContainsTask(parent_task_id))
+              << "Task metadata not found in either GCS or lineage cache. It may have been "
+                 "evicted "
+              << "by the redis LRU configuration. Consider increasing the memory "
+                 "allocation via "
+              << "ray.init(redis_max_memory=<max_memory_bytes>).";
+          // Use a copy of the cached task spec to get the parent actor id.
+          Task parent_task = lineage_cache_.GetTaskOrDie(parent_task_id);
+          new_actor_data.set_parent_actor_id(parent_task.GetTaskSpecification().ActorId().Binary());
+          FinishAssignedActorTaskHelper(actor_id, new_actor_data, resumed_from_checkpoint);
+          if(!resumed_from_checkpoint){
+            // The actor was not resumed from a checkpoint. We extend the actor's
+            // frontier as usual since there is no frontier to restore.
+            ExtendActorFrontier(task, actor_id, actor_handle_id);
+          }
+        }));
   }
-
-  if (!resumed_from_checkpoint) {
+  else if (!resumed_from_checkpoint) {
     // The actor was not resumed from a checkpoint. We extend the actor's
     // frontier as usual since there is no frontier to restore.
-    auto actor_entry = actor_registry_.find(actor_id);
-    RAY_CHECK(actor_entry != actor_registry_.end());
-    // Extend the actor's frontier to include the executed task.
-    const auto dummy_object = task.GetTaskSpecification().ActorDummyObject();
-    const ObjectID object_to_release =
-        actor_entry->second.ExtendFrontier(actor_handle_id, dummy_object);
-    if (!object_to_release.IsNil()) {
-      // If there were no new actor handles created, then no other actor task
-      // will depend on this execution dependency, so it safe to release.
-      HandleObjectMissing(object_to_release);
-    }
-    // Mark the dummy object as locally available to indicate that the actor's
-    // state has changed and the next method can run. This is not added to the
-    // object table, so the update will be invisible to both the local object
-    // manager and the other nodes.
-    // NOTE(swang): The dummy objects must be marked as local whenever
-    // ExtendFrontier is called, and vice versa, so that we can clean up the
-    // dummy objects properly in case the actor fails and needs to be
-    // reconstructed.
-    HandleObjectLocal(dummy_object);
+    ExtendActorFrontier(task, actor_id, actor_handle_id);
+  }
+}
+
+void NodeManager::ExtendActorFrontier(const Task &task, ActorID &actor_id, ActorHandleID &actor_handle_id){
```

Comment:
This looks like it only needs to take in `task.GetTaskSpecification().ActorDummyObject()`, not the whole `task`. This would allow us to avoid capturing `task` in `FinishAssignedActorTask`.

Reply:
That makes sense. However, we would still need to capture one of TaskSpecification and ActorTableData as CreateActorTableDataFromCreationTask would require TaskSpecification if we move the CreateActorTableDataFromCreationTask call to the helper function

Reply:
Takes dummy_object now instead of the entire task.